### PR TITLE
Ports Accurate donut box sprites

### PR DESCRIPTION
**Pull request by ktlwjec0**
https://github.com/tgstation/tgstation/pull/51971
 so you can see what donuts are inside and in what order

image

:cl:  Mothblocks, Ktlwjec
imageadd: Donut boxes will now accurately show what donuts are inside.
/:cl:

### CODE REVIEW

**Comment by solwashere** [comment]
Sec buff, hugbox, ided

**Comment by ktlwjec0** [comment]
is there a better way to make it apply to donut_box/deadly or should i just copypaste the code again for that too

**Review by cuackles — APPROVED** [review]

File: code/game/objects/items/storage/fancy.dm [hunk 1]
```diff
@@ -81,9 +82,38 @@
...
+/obj/item/storage/fancy/donut_box/update_overlays()
+	. = ..()
+
+	if (!fancy_open)
+		return
+
+	var/donuts = 0
+
+	for (var/_donut in contents)
+		var/obj/item/reagent_containers/food/snacks/donut/donut = _donut
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm:105]:
```suggestion
	for (var/obj/item/reagent_containers/food/snacks/donut/donut as anything in contents)
```

File: code/game/objects/items/storage/fancy.dm [hunk 2]
```diff
@@ -81,9 +82,38 @@
...
+		if (!istype(donut))
+			continue
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm]:
```suggestion
	for (var/obj/item/reagent_containers/food/snacks/donut/donut in contents)
```

The for() loop and then typepath is meant to be an optimization for things you know there will only have that one typepath in. In this case, you're expecting all contents to be donuts. However, you then check if that thing is a donut, making this whole thing pointless.

File: code/game/objects/items/storage/fancy.dm [hunk 3]
```diff
@@ -99,7 +129,36 @@
...
 
+	for (var/_donut in contents)
+		var/obj/item/reagent_containers/food/snacks/donut/donut = _donut
+		if (!istype(donut))
+			continue
+
+		. += image(icon = initial(icon), icon_state = donut.in_box_sprite(), pixel_x = donuts * DONUT_INBOX_SPRITE_WIDTH)
+		donuts += 1
+
+	. += image(icon = initial(icon), icon_state = "donutbox_top")
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm]: This should all already exist from parent, no need to copy paste all of this

Comment (ktlwjec0) [reply code/game/objects/items/storage/fancy.dm]: Done but, sprite isn't showing correctly until you remove a donut from the box, not sure how to fix that.

File: code/game/objects/items/storage/fancy.dm [hunk 4]
```diff
@@ -81,9 +82,34 @@
 	for(var/i in 1 to 6)
 		new /obj/item/reagent_containers/food/snacks/donut(src)
 
+/obj/item/storage/fancy/donut_box/PopulateContents()
+	. = ..()
+	update_icon()
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm]: At the end of ``Initialize()``, call ``update_appearance(UPDATE_ICON)`` instead

File: code/game/objects/items/storage/fancy.dm [hunk 5]
```diff
@@ -81,9 +82,34 @@
...
+	var/donuts = 0
+
+	for (var/obj/item/reagent_containers/food/snacks/donut/donut in contents)
+
+		. += image(icon = initial(icon), icon_state = donut.in_box_sprite(), pixel_x = donuts * DONUT_INBOX_SPRITE_WIDTH)
+		donuts += 1
+
+	. += image(icon = initial(icon), icon_state = "donutbox_top")
+
 /obj/item/storage/fancy/donut_box/deadly
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm:111]: There is an ``Initialize()`` call under this, but it is on donuts itself and not deadly donut box. Furthermore, it's completely useless as parent already handles it. Please remove it.

File: code/game/objects/items/storage/fancy.dm [hunk 6]
```diff
@@ -81,9 +82,34 @@
 	for(var/i in 1 to 6)
 		new /obj/item/reagent_containers/food/snacks/donut(src)
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm]:
```suggestion
		new spawn_type(src)
```
If you do this, you can remove ``PopulateContents`` from the deadly donut box too, making it also use parent (therefore harder for it to get bugs)

File: code/game/objects/items/storage/fancy.dm [hunk 7]
```diff
@@ -81,9 +82,34 @@
...
+		donuts += 1
+
+	. += image(icon = initial(icon), icon_state = "donutbox_top")
+
 /obj/item/storage/fancy/donut_box/deadly
 	icon = 'icons/obj/food/containers.dmi'
-	icon_state = "donutbox6"
 	icon_type = "donut"
 	name = "donut box"
 	spawn_type = /obj/item/reagent_containers/food/snacks/donut/deadly
```
Comment (JohnFulpWillard) [inline code/game/objects/items/storage/fancy.dm]: The only thing that the deadly donut box should have is the defines on its ``spawn_type``. Everything else is copy pasted.

**Comment by ktlwjec0** [comment]
Thank you!! This should be ready now

**Review by Bibby0110 — APPROVED** [review]

**Pull request merged**